grpc/grpc-macros/src/lib.rs

```diff
@@ -53,30 +53,66 @@ impl GrpcMethod {
+                let transports = self.inner.transports.clone();
                 let param = crate::grpc::IntoGrpcParam::into_parameter(( #( #params ),* ));
 
                 crate::grpc::AsyncGrpcCall::new(move |context: crate::grpc::Context| async move {
-                    // 256 mb, future proof as celesita blocks grow
+                    // 256 mb, future proof as celestia blocks grow
                     const MAX_MSG_SIZE: usize = 256 * 1024 * 1024;
 
-                    let mut client = #grpc_client_struct :: new(transport)
-                        .max_decoding_message_size(MAX_MSG_SIZE)
-                        .max_encoding_message_size(MAX_MSG_SIZE);
-
-                    let request = ::tonic::Request::from_parts(
-                        context.metadata,
-                        ::tonic::Extensions::new(),
-                        param,
-                    );
-
-                    let fut = client. #grpc_method_name (request);
-
-                    #[cfg(target_arch = "wasm32")]
-                    let fut = ::send_wrapper::SendWrapper::new(fut);
-
-                    let resp = fut.await?.into_inner();
-                    crate::grpc::FromGrpcResponse::try_from_response(resp)
+                    let mut last_error: Option<crate::Error> = None;
+
+                    let transport_snapshot: Vec<_> = {
+                        let guard = transports.lock().await;
+                        guard.iter().cloned().collect()
+                    };
+
+                    for (idx, transport) in transport_snapshot.into_iter().enumerate() {
+                        let transport_url = transport.metadata.url.clone();
+                        let mut client = #grpc_client_struct::new(transport)
+                            .max_decoding_message_size(MAX_MSG_SIZE)
+                            .max_encoding_message_size(MAX_MSG_SIZE);
+
+                        let mut request = ::tonic::Request::from_parts(
+                            context.metadata.clone(),
+                            ::tonic::Extensions::new(),
+                            ::std::clone::Clone::clone(&param),
+                        );
+                        request.set_timeout(::std::time::Duration::from_secs(30));
+
+                        let fut = client.#grpc_method_name(request);
+
+                        #[cfg(target_arch = "wasm32")]
+                        let fut = ::send_wrapper::SendWrapper::new(fut);
+
+                        match fut.await {
+                            Ok(resp) => {
+                                // If we succeeded on a fallback endpoint (idx > 0),
+                                // move it to the front for future requests
+                                if idx > 0 {
+                                    let mut guard = transports.lock().await;
+                                    if guard.len() > idx {
+                                        guard.swap(0, idx);
+                                    }
```

Comment:
swap here doesn't really work. We should instead make a swap on a local vector (which is a bit problematic because we iter it, maybe we need to switch to only `idx`) and just `mem::replace`. If you run 2 parallel grpc calls at the same time, and they both reach the `.lock()` here, one will just undo the work of the other.
Although since we clone here anyway, and the `Vec` only has `Box`es, maybe that's actually a good place to remove any locking and have an [`ArcSwap`](https://docs.rs/arc-swap/latest/arc_swap/type.ArcSwap.html)? @oblique wdyt

Reply:
ah ok thats exactly what you suggested